haystack/nodes/label_generator/pseudo_label_generator.py

```diff
@@ -0,0 +1,211 @@
+                    question_doc_pairs.append({"question": q.strip(), "document": documents[idx].content})
+        return question_doc_pairs
+
+    def mine_negatives(self, question_doc_pairs: List[Dict[str, str]], batch_size: int = 2) -> List[Dict[str, str]]:
+        """
+        Given a list of question and pos_doc pairs, this function returns a list of question/pos_doc/neg_doc
+        dictionaries
+
+        :param question_doc_pairs: A list of question/pos_doc pairs
+        :type question_doc_pairs: List[Dict[str, str]]
+        :param batch_size: The number of queries to run in a batch, defaults to 2
+        :type batch_size: int (optional)
+        :return: A list of dictionaries, where each dictionary contains the question, positive document,
+                and negative document.
+        """
+        question_pos_doc_neg_doc: List[Dict[str, str]] = []
+
+        for i in range(0, len(question_doc_pairs), batch_size):
+            # question in batches to minimize network latency
+            i_end = min(i + batch_size, len(question_doc_pairs))
+            queries: List[str] = [e["question"] for e in question_doc_pairs[i:i_end]]
+            pos_docs: List[str] = [e["document"] for e in question_doc_pairs[i:i_end]]
+
+            docs: List[List[Document]] = self.retriever.retrieve_batch(
+                queries=queries, top_k=self.top_k, batch_size=batch_size
+            )
+
+            # iterate through queries and find negatives
+            for question, pos_doc, top_docs in zip(queries, pos_docs, docs):
+                random.shuffle(top_docs)
+                for doc_item in top_docs:
+                    neg_doc = doc_item.content
+                    if neg_doc != pos_doc:
+                        question_pos_doc_neg_doc.append({"question": question, "pos_doc": pos_doc, "neg_doc": neg_doc})
+                        break
+        return question_pos_doc_neg_doc
+
+    def generate_margin_scores(self, mined_negatives: List[Dict[str, str]]) -> List[Dict]:
+        """
+        Given a list of mined negatives, predict the score margin between the positive and negative document using
+        the cross encoder.
+
+        The function returns a list of examples, where each example is a dictionary with the following keys:
+
+        * question: the question string
+        * pos_doc: the positive document string
+        * neg_doc: the negative document string
+        * score: the score margin
+
+        :param mined_negatives: List[Dict[str, str]] = List of mined negatives
+        :type mined_negatives: List[Dict[str, str]]
+        :return: A list of dictionaries, each of which has the following keys:
+            - question: The question string
+            - pos_doc: The positive document string
+            - neg_doc: The negative document string
+            - score: The score margin
+        """
+        examples: List[Dict] = []
+        for item in mined_negatives:
```

Comment:
As batching is possible, let's use batching here and have an Optional batch_size parameter for this method. If this parameter is not set, we can use the `batch_size` set in the init method.